File: backend/models/research.py
"""Typed models for deep research reports."""

from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field

from models.content.items import ContentItemKind
from models.conversation import SourceKind
from models.enums import ResearchRunStatus


class ResearchReportSource(BaseModel):
    """Document citation metadata attached to one research report."""

    file_path: str
    content_item_id: str | None = None
    display_name: str | None = None
    content_kind: ContentItemKind | None = None
    email_from_address: str | None = None
    email_sent_at: datetime | None = None
    parent_display_name: str | None = None
    title: str | None = None
    source_kind: SourceKind | None = None
    page_numbers: list[int] = Field(default_factory=list)
    doc_refs: list[str] = Field(default_factory=list)
    citation_index: int | None = None
    images: list[str] = Field(default_factory=list)
    quote: str | None = None


class ResearchReportImage(BaseModel):
    """Image selected for inclusion in one research report."""

    url: str
    title: str | None = None
    citation_index: int | None = None


class ResearchReportSection(BaseModel):
    """One section of the final research report."""

    heading: str
    body: str


class ResearchVerification(BaseModel):
    """Validation notes gathered during report assembly."""

    issues: list[str] = Field(default_factory=list)


class ResearchReportSummary(BaseModel):
    """List item returned for one owned research report."""

    id: str
    title: str | None
    prompt: str
    status: ResearchRunStatus
    created_at: str
    updated_at: str
    finished_at: str | None = None


class ResearchReportDetail(BaseModel):
    """Detailed research report payload returned to the frontend."""

    model_config = ConfigDict(use_enum_values=True)

    id: str
    title: str | None
    prompt: str
    status: ResearchRunStatus
    context_file_paths: list[str] = Field(default_factory=list)
    outline: list[str] = Field(default_factory=list)
    sections: list[ResearchReportSection] = Field(default_factory=list)
    sources: list[ResearchReportSource] = Field(default_factory=list)
    images: list[ResearchReportImage] = Field(default_factory=list)
    verification: ResearchVerification = Field(default_factory=ResearchVerification)
    content_markdown: str | None = None
    error_message: str | None = None
    run_id: str | None = None
    created_at: str
    updated_at: str
    finished_at: str | None = None


class ResearchReportListResponse(BaseModel):
    """Response payload for listing owned research reports."""

    reports: list[ResearchReportSummary]
    total: int

File: frontend/src/api/research.ts
import type { ContentItemKind } from "./content.ts";

export interface ResearchReportSource {
  file_path: string;
  content_item_id?: string | null;
  display_name?: string | null;
  content_kind?: ContentItemKind | null;
  email_from_address?: string | null;
  email_sent_at?: string | null;
  parent_display_name?: string | null;
  title?: string | null;
  source_kind?: "reviewed_enrichment" | null;
  page_numbers: number[];
  doc_refs: string[];
  citation_index?: number | null;
  images: string[];
  quote?: string | null;
}

export interface ResearchReportImage {
  url: string;
  title?: string | null;
  citation_index?: number | null;
}

export interface ResearchReportSection {
  heading: string;
  body: string;
}

export interface ResearchVerification {
  issues: string[];
}

export interface ResearchRun {
  id: string;
  report_id: string;
  user_sub: string;
  status: "PENDING" | "RUNNING" | "COMPLETED" | "FAILED" | "CANCELLED";
  claimed_by?: string | null;
  claimed_at?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  last_event_id?: string | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ResearchReportSummary {
  id: string;
  title: string | null;
  prompt: string;
  status: ResearchRun["status"];
  created_at: string;
  updated_at: string;
  finished_at?: string | null;
}

export interface ResearchReportDetail {
  id: string;
  title: string | null;
  prompt: string;
  status: ResearchRun["status"];
  context_file_paths: string[];
  outline: string[];
  sections: ResearchReportSection[];
  sources: ResearchReportSource[];
  images: ResearchReportImage[];
  verification: ResearchVerification;
  content_markdown?: string | null;
  error_message?: string | null;
  run_id?: string | null;
  created_at: string;
  updated_at: string;
  finished_at?: string | null;
}

export interface ResearchReportMessageMetadata extends ResearchReportDetail {
  kind: "research_report";
  report_id: string;
  conversation_id: string;
}
